### PR TITLE
Fix/prm reward

### PR DESCRIPTION
**Pull request by Ekaanksh-dev**
## Type of change

- [ ] Bug fix
- [X] New feature
- [ ] Refactor / cleanup
- [ ] Documentation
- [X] Tests

## Checklist

- [X] `ruff check src/soup_cli/ tests/` passes
- [X] `pytest tests/ -v` passes
- [ ] Updated relevant docs (`README.md` and the matching page under `docs/`) if needed

### CODE REVIEW

**Review by MakazhanAlpamys — CHANGES_REQUESTED** [review]
Thanks for taking this on, @Ekaanksh-dev — and for cleanly removing the notify workflow. One quick note on that: please keep personal CI (the `Punk Records` / `devops-ai-notifier` workflow) out of PRs to this repo — it wires up your own `GROQ_API_KEY` / email secrets that don't exist here, so on Soup it would just add failing runs. You already deleted it so the final diff is clean — just flagging for next time. 🙏

On the change itself, I think it needs one more pass — the core goal of #298 isn't met yet:

**1. The forward pass is still per-completion (the main thing).** The new code builds a padded `[B, T]` tensor + attention mask (good), but then loops row-by-row and forwards one row at a time:

```python
for b, orig_idx in enumerate(valid_indices):
    single_ids = padded[b:b+1]             # [1, T]
    ...
    outputs = self._model(**model_kwargs)  # one forward per completion
```

So it's still `len(valid)` forward passes — same count as before — but now each is padded to `max_len`, i.e. *more* compute than the original unpadded path. #298 asks for a **single** `output_hidden_states` forward over the whole `[B, T]` batch. Hoist the model call out of the loop:

```python
outputs = self._model(input_ids=padded, attention_mask=attn_mask, output_hidden_states=True)
last_hidden = outputs.hidden_states[-1]           # [B, T, H]
for b, orig_idx in enumerate(valid_indices):
    pos = torch.tensor(step_positions_for_row_b, device=self.device)   # all < row_len <= T
    step_hidden = last_hidden[b].index_select(0, pos)
    scores = self._model.reward_head(step_hidden).squeeze(-1)
    ...
```

Because every `step_position` is `< row_len <= max_len` and you right-pad, the boundary hidden states are identical to the unbatched result once the attention mask is applied.

**2. Needs the parity test (AC #1).** The PR only touches `prm_reward.py` — no test. The acceptance criteria ask for a unit test using the injected fake-model fixture already in `tests/test_v07130.py` (`_FakeModel`, `hidden_states[-1] = arange(T)`) asserting the batched path returns **numerically-equal** rewards to the per-completion loop, plus a **mixed-length** case (a short and a long completion in the same call) to prove the padding mask prevents cross-row contamination. That mixed-length case is the crux — it's the test that would catch a batching bug.

**3. Minor, once the above lands:**
- `inspect.signature(self._model.__call__)` inside the loop is fragile + per-iteration; with a single batched call you can pass `attention_mask` unconditionally (real HF models accept it).
- Keep the existing `_score_one` per-completion path as the fallback / parity reference (the issue asks for this) — convenient to assert against in the new test.

The overall shape (pad -> mask -> gather boundaries) is exactly right — it just needs the single forward + the test. Ping me when it's updated and I'll re-review. Thanks for digging in! 🙏

**Comment by Ekaanksh-dev** [comment]
@MakazhanAlpamys sorry for the Punk Records / devops-ai-notifier workflow I am fixing that .I have solve the pass you stated

**Comment by MakazhanAlpamys** [comment]
Merged — thank you @Ekaanksh-dev! 🎉

This nails #298: `PRMScorer.__call__` now builds one right-padded `[B, T]` batch (+ attention mask) and runs a **single** `output_hidden_states` forward, then gathers each row's step-boundary hidden states — instead of one forward per completion. The per-completion `_score_one` stays as the parity reference, exactly as the issue asked.

Verified on your exact commit: the two new tests — `test_batched_parity_with_per_completion` and `test_mixed_length_no_cross_row_contamination` — are the right ones (the mixed-length case is what proves the padding mask doesn't leak across rows), and `test_v07130.py` passes 16/16 locally, full CI matrix green. I pushed one tiny follow-up (`B` -> `batch` to satisfy ruff N806) so lint went green — no logic change. Credited you in [CONTRIBUTORS.md](https://github.com/MakazhanAlpamys/Soup/blob/main/CONTRIBUTORS.md) + the CHANGELOG.

Really nice turnaround incorporating the review feedback — and no worries at all about the earlier workflow, it's fully sorted. Hope to see more PRs from you! 🙏
